[WIP] Add tests for MapDataset C++ kernel

feihugis opened the pull request:
This PR adds the tests for the public APIs of MapDataset C++ kernel.

Besides, it adds class `DatasetOpsTestBase` to provide the following utilities for testing Dataset kernels:

- prepare `TreadPool` and `FunctionLibraryRuntime`.

- make `OpKernelContext`, `OpKernel`, and run `OpKernel`.

- add inputs for `OpKernel` and get the output dataset from `OpKernelContext`.

- make `Dataset`, `Iterator`, and `IteratorContext`, and provide `GetNext`.

Besides, these API designs enable the subclass to own multiple kernels and datasets, which are necessary to run the tests for some `Dataset`s (e.g. MapDataset).

feihugis commented:
@jsimsa This PR is still working in progress. The current one is to update the status of the PR with you, not ready for review yet. 

The tests for RangeDataset work fine. However, when running  MapDataset tes, it often throws the following error. 
```
F tensorflow/core/common_runtime/scoped_allocator_mgr.cc:81] Failed to find instance 1073743869 in container 0 on /job:a/replica:0/task:0/device:CPU:0
*** Received signal 6 ***
*** BEGIN MANGLED STACK TRACE ***
0   libtensorflow_framework.so          0x000000011d93c4c4 _ZN10tensorflow7testing19SafePrintStackTraceEv + 84
1   libtensorflow_framework.so          0x000000011d93c24e _ZN10tensorflow7testingL17StacktraceHandlerEiP9__siginfoPv + 318
2   libsystem_platform.dylib            0x00007fff7d3c6f5a _sigtramp + 26
3   ???                                 0x0000000000000000 0x0 + 0
4   libsystem_c.dylib                   0x00007fff7d1641ae abort + 127
5   libtensorflow_framework.so          0x000000011df20dba _ZN10tensorflow8internal15LogMessageFatalD2Ev + 58
6   libtensorflow_framework.so          0x000000011df20ddc _ZN10tensorflow8internal15LogMessageFatalD1Ev + 28
7   libtensorflow_framework.so          0x000000011e3cbfee _ZN10tensorflow24ScopedAllocatorContainer11GetInstanceEi + 1758
8   libtensorflow_framework.so          0x000000011d82aea8 _ZN10tensorflow16ThreadPoolDevice18GetScopedAllocatorENS_19AllocatorAttributesEx + 104
9   libtensorflow_framework.so          0x000000011d070923 _ZN10tensorflow15OpKernelContext13get_allocatorENS_19AllocatorAttributesE + 147
10  libtensorflow_framework.so          0x000000011d076a78 _ZN10tensorflow15OpKernelContext15allocate_tensorENS_8DataTypeERKNS_11TensorShapeEPNS_6TensorENS_19AllocatorAttributesERKNS_20AllocationAttributesE + 104
11  libtensorflow_framework.so          0x000000011d077183 _ZN10tensorflow15OpKernelContext15allocate_tensorENS_8DataTypeERKNS_11TensorShapeEPNS_6TensorENS_19AllocatorAttributesE + 163
12  libtensorflow_framework.so          0x000000011d076584 _ZN10tensorflow15OpKernelContext15allocate_outputEiRKNS_11TensorShapeEPPNS_6TensorENS_19AllocatorAttributesE + 980
13  libtensorflow_framework.so          0x000000011d076196 _ZN10tensorflow15OpKernelContext15allocate_outputEiRKNS_11TensorShapeEPPNS_6TensorE + 646
14  libtensorflow_framework.so          0x000000011cf183ec _ZN10tensorflow4data15DatasetOpKernel7ComputeEPNS_15OpKernelContextE + 220
15  libtensorflow_framework.so          0x000000011d696d1b _ZN10tensorflow6Device7ComputeEPNS_8OpKernelEPNS_15OpKernelContextE + 59
16  map_dataset_op_test                 0x0000000106db5164 _ZN10tensorflow4data18DatasetOpsTestBase11RunOpKernelEPNS_8OpKernelEPNS_15OpKernelContextE + 100
17  map_dataset_op_test                 0x0000000106d62a0e _ZN10tensorflow4data12_GLOBAL__N_116MapDatasetOpTest11MakeDatasetEPNS0_11DatasetBaseEi + 1646
18  map_dataset_op_test                 0x0000000106d7d79b _ZN10tensorflow4data12_GLOBAL__N_142MapDatasetOpTest_IteratorOutputShapes_Test8TestBodyEv + 2699
19  map_dataset_op_test                 0x0000000107537bfe _ZN7testing8internal38HandleSehExceptionsInMethodIfSupportedINS_4TestEvEET0_PT_MS4_FS3_vEPKc + 126
20  map_dataset_op_test                 0x00000001075032cb _ZN7testing8internal35HandleExceptionsInMethodIfSupportedINS_4TestEvEET0_PT_MS4_FS3_vEPKc + 123
21  map_dataset_op_test                 0x0000000107503203 _ZN7testing4Test3RunEv + 211
22  map_dataset_op_test                 0x0000000107504c9f _ZN7testing8TestInfo3RunEv + 223
23  map_dataset_op_test                 0x000000010750667c _ZN7testing8TestCase3RunEv + 236
24  map_dataset_op_test                 0x00000001075206b9 _ZN7testing8internal12UnitTestImpl11RunAllTestsEv + 905
25  map_dataset_op_test                 0x000000010753ab6e _ZN7testing8internal38HandleSehExceptionsInMethodIfSupportedINS0_12UnitTestImplEbEET0_PT_MS4_FS3_vEPKc + 126
26  map_dataset_op_test                 0x00000001075200bb _ZN7testing8internal35HandleExceptionsInMethodIfSupportedINS0_12UnitTestImplEbEET0_PT_MS4_FS3_vEPKc + 123
27  map_dataset_op_test                 0x000000010751ff7c _ZN7testing8UnitTest3RunEv + 412
28  map_dataset_op_test                 0x00000001074b2421 _Z13RUN_ALL_TESTSv + 17
29  map_dataset_op_test                 0x00000001074b2400 main + 224
30  libdyld.dylib                       0x00007fff7d0b8015 start + 1
*** END MANGLED STACK TRACE ***

*** Begin stack trace ***
	tensorflow::CurrentStackTrace()
	tensorflow::testing::StacktraceHandler(int, __siginfo*, void*)
	_sigtramp
	
	abort
	tensorflow::internal::LogMessageFatal::~LogMessageFatal()
	tensorflow::internal::LogMessageFatal::~LogMessageFatal()
	tensorflow::ScopedAllocatorContainer::GetInstance(int)
	tensorflow::ThreadPoolDevice::GetScopedAllocator(tensorflow::AllocatorAttributes, long long)
	tensorflow::OpKernelContext::get_allocator(tensorflow::AllocatorAttributes)
	tensorflow::OpKernelContext::allocate_tensor(tensorflow::DataType, tensorflow::TensorShape const&, tensorflow::Tensor*, tensorflow::AllocatorAttributes, tensorflow::AllocationAttributes const&)
	tensorflow::OpKernelContext::allocate_tensor(tensorflow::DataType, tensorflow::TensorShape const&, tensorflow::Tensor*, tensorflow::AllocatorAttributes)
	tensorflow::OpKernelContext::allocate_output(int, tensorflow::TensorShape const&, tensorflow::Tensor**, tensorflow::AllocatorAttributes)
	tensorflow::OpKernelContext::allocate_output(int, tensorflow::TensorShape const&, tensorflow::Tensor**)
	tensorflow::data::DatasetOpKernel::Compute(tensorflow::OpKernelContext*)
	tensorflow::Device::Compute(tensorflow::OpKernel*, tensorflow::OpKernelContext*)
	tensorflow::data::DatasetOpsTestBase::RunOpKernel(tensorflow::OpKernel*, tensorflow::OpKernelContext*)
	tensorflow::data::(anonymous namespace)::MapDatasetOpTest::MakeDataset(tensorflow::data::DatasetBase*, int)
	tensorflow::data::(anonymous namespace)::MapDatasetOpTest_IteratorOutputShapes_Test::TestBody()
	void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)
	void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)
	testing::Test::Run()
	testing::TestInfo::Run()
	testing::TestCase::Run()
	testing::internal::UnitTestImpl::RunAllTests()
	bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)
	bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)
	testing::UnitTest::Run()
	RUN_ALL_TESTS()
	main
	start
*** End stack trace ***
Abort trap: 6
```
I'm debugging this issue. If you have any suggestion, it will be very welcome.

jsimsa reviewed (commented):
Can you remove everything related to GPU / CUDA from your PR and see if the allocator problem goes away?

I think that your unit tests should aim to be CPU-only (most tf.data transformations do not have GPU kernels).

feihugis commented:
@jsimsa The GPU/CUDA related parts have been removed. The error is caused by some programming issues when updating `params_->output_attr_array`. Now the issue has been fixed. The tests for `MapDataset` and `RangeDataset` can successfully run on my laptop. Could you help review this PR when you have time?

jsimsa commented:
Thanks, I plan to take a look later today.

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.cc:
```diff
@@ -0,0 +1,216 @@
+limitations under the License.
+==============================================================================*/
+
+#ifdef GOOGLE_CUDA
```
get rid of this

jsimsa replied:
get rid of this

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.cc:
```diff
@@ -0,0 +1,216 @@
+#include "tensorflow/core/common_runtime/gpu/gpu_managed_allocator.h"
+#endif
+
+#include "dataset_test_base.h"
```
use a fully qualified path (starting with `tensorflow/core/...`)

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.h:
```diff
@@ -0,0 +1,129 @@
+limitations under the License.
+==============================================================================*/
+
+#ifndef TENSORFLOW_DATASET_TEST_BASE_H
```
use a fully qualified path `TENSORFLOW_CORE_KERNELS_DATASET_TEST_BASE_H_

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.h:
```diff
@@ -0,0 +1,129 @@
+    gtl::STLDeleteElements(&managed_outputs_);
+  }
+
+  Status InitThreadPool(int thread_num);
```
all public method should have documentation that briefly explains what does the method do, including what are the arguments used for.

jsimsa replied:
in addition, arguments that are not modified by the method, should be passed in as a constant reference (e.g. `const OpKernel& kernel`)

jsimsa replied:
also, do any of these method need to be public or can be protected?

feihugis committed:
Add tests for MapDataset C++ kernel

jsimsa commented on tensorflow/core/kernels/data/map_dataset_op_test.cc:
```diff
@@ -0,0 +1,384 @@
+    range_dataset_->Unref();
+  }
+
+ public:
```
no need for `public` here ... I think everything below should be either `protected` or `private`

jsimsa commented on tensorflow/core/kernels/data/map_dataset_op_test.cc:
```diff
@@ -0,0 +1,384 @@
+  FunctionDef MakeFuncDef() { return test::function::XTimesTwo(); }
+
+  string FuncName() { return "XTimesTwo"; }
+  template <typename T>
```
add empty line

jsimsa commented on tensorflow/core/kernels/data/map_dataset_op_test.cc:
```diff
@@ -0,0 +1,384 @@
+    return Status::OK();
+  }
+
+ protected:
```
private?

jsimsa replied:
ping

feihugis replied:
If changing `protected` to `private`, these private members can not be accessed in `TEST_P` or `TEST_F`.

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.h:
```diff
@@ -0,0 +1,129 @@
+  Status GetNext(IteratorBase* iterator, IteratorContext* iterator_context,
+                 std::vector<Tensor>* out_tensors, bool* end_of_sequence);
+
+  Allocator* allocator();
```
I don't think you need this method, the derived classes can just use `allocator_`.

jsimsa commented on tensorflow/core/kernels/data/map_dataset_op_test.cc:
```diff
@@ -0,0 +1,384 @@
+
+  string FuncName() { return "XTimesTwo"; }
+  template <typename T>
+  Status MakeRangeDatasetOpKernel() {
```
instead of using internal state, have the function return the result in an output argument

I also think that this will be used commonly by other unit test and makes sense to put it into the test base

jsimsa commented on tensorflow/core/kernels/data/map_dataset_op_test.cc:
```diff
@@ -0,0 +1,384 @@
+    return Status::OK();
+  }
+
+  Status MakeRangeDataset(int start, int end, int step, int output_index) {
```
again, the method should return the dataset in an output argument and it should go to the test base

jsimsa commented on tensorflow/core/kernels/data/map_dataset_op_test.cc:
```diff
@@ -0,0 +1,384 @@
+  }
+
+  template <typename T>
+  Status MakeMapDatasetOpKernel() {
```
ditto

feihugis replied:
@jsimsa Besides returning the dataset as an output argument, do you also mean `MakeMapDatasetOpKernel ` and `MakeMapDatasetOpKernel ` need to be moved to the test base? I just want to confirm it.

jsimsa replied:
sorry for the confusion, just returning the result by output argument but keeping it here for the time being

feihugis replied:
Got it. Thanks.

jsimsa commented on tensorflow/core/kernels/data/map_dataset_op_test.cc:
```diff
@@ -0,0 +1,384 @@
+    return Status::OK();
+  }
+
+  Status MakeMapDataset(DatasetBase* input_dataset, int output_index) {
```
ditto

jsimsa commented on tensorflow/core/kernels/data/map_dataset_op_test.cc:
```diff
@@ -0,0 +1,384 @@
+  int step = std::get<2>(GetParam());
+  int breakpoint = std::get<3>(GetParam());
+  TF_ASSERT_OK(InitThreadPool(thread_num));
+  TF_ASSERT_OK(InitFunctionLibraryRuntime({MakeFuncDef()}, cpu_num));
```
Can we test with a couple of different functions?

jsimsa commented on tensorflow/core/kernels/data/map_dataset_op_test.cc:
```diff
@@ -0,0 +1,384 @@
+  }
+
+ public:
+  FunctionDef MakeFuncDef() { return test::function::XTimesTwo(); }
```
I think introducing this function just adds unnecessary indirection.

jsimsa commented on tensorflow/core/kernels/data/map_dataset_op_test.cc:
```diff
@@ -0,0 +1,384 @@
+  int end = std::get<1>(GetParam());
+  int step = std::get<2>(GetParam());
+  TF_ASSERT_OK(InitThreadPool(thread_num));
+  TF_ASSERT_OK(InitFunctionLibraryRuntime({MakeFuncDef()}, cpu_num));
```
can we test with a couple of different function (using test parametrization)

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.h:
```diff
@@ -0,0 +1,129 @@
+                          DataTypeVector input_types, DataType dtype,
+                          const TensorShape& shape);
+
+  Status MakeDataset(OpKernel* kernel, OpKernelContext* context);
```
this method should return `DatasetBase*` through an output argument instead of updating internal state

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.h:
```diff
@@ -0,0 +1,129 @@
+
+  Status MakeDataset(OpKernel* kernel, OpKernelContext* context);
+
+  Status MakeIterator(OpKernelContext* context, DatasetBase* dataset,
```
ditto, this should return `IteratorBase*` through an output argument

jsimsa reviewed (changes requested):
@feihugis thanks, did a first rough pass a left a bunch of comments ... one of the main repeated comments is that methods that create objects used by the tests should return these objects instead of stored them in an internal state ... this will result in more intuitive API

feihugis committed:
Refactor the test base and related tests:

- Add the comments for all the public methods.

- Change the methods to return objects instead of storing them in an internal state.

- Reorganize the methods into public, protected, and private.

- Add the tests of different functions using test parameterization.

feihugis commented:
@jsimsa Thanks for your review! This [commit](https://github.com/feihugis/tensorflow/pull/4/commits/8b0398c4a04e689e3ced1e4b4fdb4a3e9f4df02f) revises the PR based on your suggestions. Could you have a look at the changes?

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.cc:
```diff
@@ -0,0 +1,182 @@
+  std::unique_ptr<OpKernel> kernel =
+      tensorflow::CreateOpKernel(device_type_, device_.get(), allocator_,
+                                 node_def, TF_GRAPH_DEF_VERSION, &status);
+  TF_CHECK_OK(status) << status;
```
The `Status` should be propagated to the caller. To do that, change the API to be:

```
Status DatasetOpsTestBase::CreateOpKernel(
    const NodeDef& node_def, std::unique_ptr<OpKernel>* op_kernel);
```

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.cc:
```diff
@@ -0,0 +1,182 @@
+
+DatasetBase* DatasetOpsTestBase::CreateDataset(OpKernel* kernel,
+                                               OpKernelContext* context) {
+  TF_CHECK_OK(RunOpKernel(kernel, context));
```
The `Status` should be propagated to the caller. To do that, change the API to be:

```
Status DatasetOpsTestBase::CreateDataset(
    OpKernelContext* context, OpKernel* kernel, DatasetBase** const dataset);
```

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.cc:
```diff
@@ -0,0 +1,182 @@
+  return GetDatasetFromContext(context, 0);
+}
+
+std::unique_ptr<IteratorBase> DatasetOpsTestBase::CreateIterator(
```
This method should not be returning both IteratorContext and IteratorBase. Also, it should be propagating the Status to the caller.

Assuming the `IteratorContext` is created from `OpKernelContext` by the caller, the API should be:

```
Status DatasetOpsTestBase::CreateIterator(
    IteratorContext* context, const DatasetBase* dataset,
    std::unique_ptr<IteratorBase>* iterator);
```

feihugis replied:
Yes, have added another method to create the IteratorContext:
```
Status CreateIteratorContext(
      OpKernelContext* const op_context,
      std::unique_ptr<IteratorContext>* iterator_context);
```

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.cc:
```diff
@@ -0,0 +1,182 @@
+  return iterator->GetNext(iterator_context, out_tensors, end_of_sequence);
+}
+
+DatasetBase* DatasetOpsTestBase::GetDatasetFromContext(OpKernelContext* context,
```
```
Status DatasetOpsTestBase::GetDatasetFromContext(OpKernelContext* context, int output_index, DatasetBase** const dataset);
```

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.cc:
```diff
@@ -0,0 +1,182 @@
+std::unique_ptr<IteratorBase> DatasetOpsTestBase::CreateIterator(
+    OpKernelContext* context, DatasetBase* dataset,
+    std::unique_ptr<IteratorContext>* iterator_context) {
+  *iterator_context = MakeUnique<IteratorContext>(context);
```
replace `MakeUnique` with `absl::make_unique` throughout this PR

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.cc:
```diff
@@ -0,0 +1,182 @@
+    std::unique_ptr<IteratorContext>* iterator_context) {
+  *iterator_context = MakeUnique<IteratorContext>(context);
+  IteratorContext::Params params(iterator_context->get());
+  params.function_handle_cache = new FunctionHandleCache(flr_);
```
where is this object deleted?

jsimsa replied:
the object should be either owned by the caller or the instance of the class ... as it is now, it is leaked

feihugis replied:
Add a member into the class: `std::unique_ptr<FunctionHandleCache> function_handle_cache_;`

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.cc:
```diff
@@ -0,0 +1,182 @@
+  *iterator_context = MakeUnique<IteratorContext>(context);
+  IteratorContext::Params params(iterator_context->get());
+  params.function_handle_cache = new FunctionHandleCache(flr_);
+  iterator_context->reset(new IteratorContext(params));
```
you should use `absl::make_unique` here

Also, you do not need to set the `IteratorContext` twice, instead you can do:

```
IteratorContext::params params(context);
... update parameters ...
 *iterator_context = absl::make_unique<IteratorContext>(std::move(params));
```

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.cc:
```diff
@@ -0,0 +1,182 @@
+  return context->status();
+}
+
+std::unique_ptr<OpKernelContext> DatasetOpsTestBase::CreateOpKernelContext(
```
Although this function can never fail, I would still change its signature to return `Status` to make it possible to add something to the function body that could fail.

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.cc:
```diff
@@ -0,0 +1,182 @@
+}
+
+std::unique_ptr<SerializationContext>
+DatasetOpsTestBase::CreateSerializationContext() {
```
ditto, change the signature to return `Status`

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.cc:
```diff
@@ -0,0 +1,182 @@
+Tensor* DatasetOpsTestBase::AddDatasetInput(
+    gtl::InlinedVector<TensorValue, 4>* inputs, DataTypeVector input_types,
+    DataType dtype, const TensorShape& shape) {
+  CHECK_GT(input_types.size(), inputs->size())
```
The method should return `Status` and this CHECK (which would crash the test) should be replaced with `return errors::InvalidArgument(...)`.

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.cc:
```diff
@@ -0,0 +1,182 @@
+  return input;
+}
+
+void DatasetOpsTestBase::SetOutputAttrs() {
```
do we need this to be a method in the API? (could we just inline the logic into the `CreateOpKernelContext` body)

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.h:
```diff
@@ -0,0 +1,191 @@
+  DatasetOpsTestBase()
+      : device_(DeviceFactory::NewDevice("CPU", {}, "/job:a/replica:0/task:0")),
+        device_type_(DEVICE_CPU) {
+    CHECK(device_.get()) << "Could not create CPU device";
```
remove

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.h:
```diff
@@ -0,0 +1,191 @@
+class DatasetOpsTestBase : public ::testing::Test {
+ public:
+  DatasetOpsTestBase()
+      : device_(DeviceFactory::NewDevice("CPU", {}, "/job:a/replica:0/task:0")),
```
replace `job:a` with `job:0`

feihugis replied:
@jsimsa `job:0` will cause the following error: `tensorflow/core/common_runtime/device.cc:28] Check failed: DeviceNameUtils::ParseFullName(name(), &parsed_name_) Invalid device name: /job:0/replica:0/task:0/device:CPU:0`. It looks like the job name needs to be alpha as indicate [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/util/device_name_utils.cc#L44).

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.h:
```diff
@@ -0,0 +1,191 @@
+ protected:
+  std::unique_ptr<Device> device_;
+  DeviceType device_type_;
+  Allocator* allocator_;
```
who owns this?

feihugis replied:
@jsimsa IIUC, `allocator_` is not owned by this class. It seems to be owned by [this class](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/allocator_registry.h#L97).

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.h:
```diff
@@ -0,0 +1,191 @@
+      : device_(DeviceFactory::NewDevice("CPU", {}, "/job:a/replica:0/task:0")),
+        device_type_(DEVICE_CPU) {
+    CHECK(device_.get()) << "Could not create CPU device";
+    allocator_ = device_->GetAllocator(AllocatorAttributes());
```
how are these allocator attributes related to `allocator_attributes_`?

feihugis replied:
`allocator_attrs_` is for allocating memory for the outputs of the op kernel and determined by `op_kernel->output_memory_types` as following:
```
const bool on_host =
        (params_->op_kernel->output_memory_types()[index] == HOST_MEMORY);
attr.set_on_host(on_host);
```
`AllocatorAttributes()` uses the default value (e.g. using cpu device memory). The derived `allocator_` is used to create the op kernel and its input tensors. 

Is my understand right?

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.h:
```diff
@@ -0,0 +1,191 @@
+  Status GetNext(IteratorBase* iterator, IteratorContext* iterator_context,
+                 std::vector<Tensor>* out_tensors, bool* end_of_sequence);
+
+  // Creates a new range dataset operation.
```
1) change docstring to `Creates a new RangeDataset op kernel`
2) change signature to propagate status to the caller
3) document the template type

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.h:
```diff
@@ -0,0 +1,191 @@
+    return CreateOpKernel(node_def);
+  }
+
+  // Creates a new range dataset.
```
1) change docstring to `Creates a new RangeDataset dataset`
2) change signature to propagate status to the caller
3) document the template type

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.h:
```diff
@@ -0,0 +1,191 @@
+  // Adds an arrayslice of data into the input vector. `input_types` describes
+  // the required data type for each input tensor. `shape` and `data` describes
+  // the shape and values of the current input tensor.
+  template <typename T>
```
document the template type

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.h:
```diff
@@ -0,0 +1,191 @@
+  std::vector<AllocatorAttributes> allocator_attrs_;
+  std::unique_ptr<ScopedStepContainer> step_container_;
+
+  FunctionLibraryRuntime* flr_;  // not owned
```
document who owns this

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.h:
```diff
@@ -0,0 +1,191 @@
+  std::unique_ptr<OpKernelContext::Params> params_;
+  std::unique_ptr<ProcessFunctionLibraryRuntime> pflr_;
+  std::unique_ptr<thread::ThreadPool> thread_pool_;
+  std::vector<Tensor*> tensors_;  // owns Tensors.
```
IIUC, this field only exists to deallocated the Tensors ... why are the Tensors created by as dynamic objects instead of values (which would be deallocated when they go out of scope)?

feihugis replied:
Yeah, `tensors_ ` is used to deallocate the tensors as below. We don't want these tensors to be released when they go out of scope, because they will be used as the inputs of the op kernel in the next step.
```C++
gtl::InlinedVector<TensorValue, 4> inputs_;
...
Tensor* input = new Tensor(allocator_, dtype, shape);
tensors_.push_back(input);
inputs_->push_back({nullptr, input});
...
params_->inputs = inputs_;
...
gtl::STLDeleteElements(&tensors_);
```

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.h:
```diff
@@ -0,0 +1,191 @@
+                                  StringPiece node_name) {
+    std::unique_ptr<OpKernel> range_kernel =
+        CreateRangeDatasetOpKernel<U>(node_name);
+    std::unique_ptr<OpKernelContext> range_context;
```
declare and define this in the same place

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.h:
```diff
@@ -0,0 +1,191 @@
+  std::unique_ptr<thread::ThreadPool> thread_pool_;
+  std::vector<Tensor*> tensors_;  // owns Tensors.
+  // Copies of the outputs in unified memory (host and device accessible).
+  std::vector<Tensor*> managed_outputs_;
```
nothing seems to be using this

feihugis replied:
Yes, I will remove it.

jsimsa reviewed (changes requested):
@feihugis thanks for the iteration ... did another pass and left some comments

feihugis committed:
Address the comments
- propagate Status to the caller
- replace MakeUnique with absl::make_unique
- memory leak issue
- documentation issue
- others

feihugis commented:
@jsimsa Thanks for your time reviewing this PR! Your comments are addressed in [this commit](https://github.com/feihugis/tensorflow/pull/4/commits/f45f56f3f80f1cfc9592bb0f0038f5b9c89b9bb7). Could you help review the changes?

BTW, do we also need to change the methods [CreateMapDataset](https://github.com/feihugis/tensorflow/pull/4/files#diff-a9d5a15b8d3bed309e32e801715f4a1dR50) and [CreateRangeDataset](https://github.com/feihugis/tensorflow/pull/4/files#diff-4085eb633f02b68967febffed95bc6edR49) to propagate `Status`?

feihugis commented on tensorflow/core/kernels/data/dataset_test_base.cc:
```diff
@@ -0,0 +1,201 @@
+  return Status::OK();
+}
+
+inline void DatasetOpsTestBase::SetOutputAttrs() {
```
@jsimsa `inline` is added here. Or we need to delete this function and move the logic into the `CreateOpKernelContext` body?

jsimsa replied:
I would just move the logic into the body of `CreateOpKernelContext`.

feihugis commented on tensorflow/core/kernels/data/dataset_test_base.h:
```diff
@@ -0,0 +1,208 @@
+ protected:
+  std::unique_ptr<Device> device_;
+  DeviceType device_type_;
+  Allocator* allocator_;  // Owned by `AllocatorFactoryRegistry`.
```
IIUC, `allocator_` is not owned by this class. It seems to be owned by [AllocatorFactoryRegistry](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/allocator_registry.h#L97).

feihugis replied:
@jsimsa `allocator_` is created [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/allocator_registry.cc#L81).

feihugis commented on tensorflow/core/kernels/data/dataset_test_base.h:
```diff
@@ -0,0 +1,208 @@
+  std::unique_ptr<FunctionLibraryDefinition> lib_def_;
+  std::unique_ptr<OpKernelContext::Params> params_;
+  std::unique_ptr<thread::ThreadPool> thread_pool_;
+  std::vector<Tensor*> tensors_;  // Owns Tensors.
```
> IIUC, this field only exists to deallocated the Tensors ... why are the Tensors created by as dynamic objects instead of values (which would be deallocated when they go out of scope)?


Yeah, `tensors_ ` is used to deallocate the tensors as below. We don't want these tensors to be released when they go out of scope, because they will be used as the inputs of the op kernel in the next step.
```C++
gtl::InlinedVector<TensorValue, 4> inputs_;
...
Tensor* input = new Tensor(allocator_, dtype, shape);
tensors_.push_back(input);
inputs_->push_back({nullptr, input});
...
params_->inputs = inputs_;
...
gtl::STLDeleteElements(&tensors_);
```

jsimsa replied:
OK. Change the type of the object to `std::vector<std::unique_ptr<Tensor>>` so that you the tensors are delete automatically when the instance of this object is deleted (as opposed to having to manually delete them in the destructor).

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.cc:
```diff
@@ -0,0 +1,201 @@
+                                         OpKernelContext* context,
+                                         DatasetBase** const dataset) {
+  TF_RETURN_IF_ERROR(RunOpKernel(kernel, context));
+  // Assume that DatasetOp has only one output.
```
Could we add a DCHECK to check this assumption?

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.cc:
```diff
@@ -0,0 +1,201 @@
+  return Status::OK();
+}
+
+Status DatasetOpsTestBase::CreateIterator(
```
After the refactor, this method does not seem to be adding much value so I would simply get rid of it.

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.cc:
```diff
@@ -0,0 +1,201 @@
+  return dataset->MakeIterator(context, "Iterator", iterator);
+}
+
+Status DatasetOpsTestBase::GetNext(IteratorBase* iterator,
```
Same comment here, I would get rid of this method.

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.cc:
```diff
@@ -0,0 +1,201 @@
+      absl::make_unique<ScopedStepContainer>(0, [](const string&) {});
+  params_->step_container = step_container_.get();
+  checkpoint::TensorSliceReaderCacheWrapper slice_reader_cache_wrapper;
+  params_->slice_reader_cache = &slice_reader_cache_wrapper;
```
you are taking an address of a local variable here, so the pointer will be invalid after this function returns

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.cc:
```diff
@@ -0,0 +1,201 @@
+  }
+  bool is_ref = IsRefType(input_types[inputs->size()]);
+  Tensor* input = new Tensor(allocator_, dtype, shape);
+  tensors_.push_back(input);
```
add a TODO here to figure out how to avoid using a member variable to garbage collect the inputs

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.h:
```diff
@@ -0,0 +1,208 @@
+
+  ~DatasetOpsTestBase() {
+    gtl::STLDeleteElements(&tensors_);
+    function_handle_cache_.reset();
```
there is no need to reset `unique_ptr` its value will be deleted when the class is destructed

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.h:
```diff
@@ -0,0 +1,208 @@
+  Status CreateOpKernel(const NodeDef& node_def,
+                        std::unique_ptr<OpKernel>* op_kernel);
+
+  // Creates a new dataset. Here we assume that the dataset operation has only
```
I would remove the "Here we assume ..." from the docstring. It is an implementation detail.

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.h:
```diff
@@ -0,0 +1,208 @@
+    return Status::OK();
+  }
+
+  // Creates a new RangeDataset dataset. `U` specifies the output dtype of the
```
use `T` instead of `U`

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.h:
```diff
@@ -0,0 +1,208 @@
+  Status InitFunctionLibraryRuntime(const std::vector<FunctionDef>& flib,
+                                    int cpu_num);
+
+  // Runs an operation producing outputs. The `context` need to be initialized
```
Remove the "The context ..." sentence.

jsimsa commented on tensorflow/core/kernels/data/map_dataset_op_test.cc:
```diff
@@ -0,0 +1,363 @@
+==============================================================================*/
+
+#include <tuple>
+#include "tensorflow/core/framework/dataset.h"
```
add an empty line here (after the `<tuple>` line)

feihugis replied:
Thanks. As the named struct is used to replace the tuple for the test parameters, this include is not needed now. I will remove it.

jsimsa commented on tensorflow/core/kernels/data/map_dataset_op_test.cc:
```diff
@@ -0,0 +1,363 @@
+ protected:
+  // Creates a new MapDataset dataset, meanwhile initializing the MapDataset op
+  // kernel and context internally.
+  DatasetBase* CreateMapDataset(DatasetBase* input_dataset,
```
for consistency, change the signature to return `Status`, replacing TF_CHECK_OK with TF_RETURN_IF_ERROR

jsimsa replied:
also, the created dataset should be returned through an output argument and not stored in the instance

jsimsa commented on tensorflow/core/kernels/data/map_dataset_op_test.cc:
```diff
@@ -0,0 +1,363 @@
+  // same with the node name of the input dataset for the method
+  // `CreateMapDataset()`.
+  template <typename T>
+  std::unique_ptr<OpKernel> CreateMapDatasetOpKernel(
```
ditto

jsimsa commented on tensorflow/core/kernels/data/map_dataset_op_test.cc:
```diff
@@ -0,0 +1,363 @@
+
+ protected:
+  DatasetBase* range_dataset_;
+  DatasetBase* map_dataset_;
```
where is this Unrefed?

jsimsa commented on tensorflow/core/kernels/data/map_dataset_op_test.cc:
```diff
@@ -0,0 +1,363 @@
+struct DatasetGetNextTest
+    : MapDatasetOpTest,
+      ::testing::WithParamInterface<
+          std::tuple<int64, int64, int64, string, std::vector<int64>,
```
Instead of using a tuple, create a struct to represent the test case. It will make the code easier to read.

jsimsa commented on tensorflow/core/kernels/data/map_dataset_op_test.cc:
```diff
@@ -0,0 +1,363 @@
+  TF_ASSERT_OK(InitFunctionLibraryRuntime(func_lib, cpu_num));
+  TF_ASSERT_OK(
+      CreateRangeDataset<int64>(start, end, step, "range", &range_dataset_));
+  map_dataset_ = CreateMapDataset(range_dataset_, func_name);
```
After changing the `CreateRangeDataset and `CreateMapDataset` signatures, this should look as follows:

```
DatasetBase* range_dataset;
TF_ASSERT_OK(
      CreateRangeDataset<int64>(start, end, step, "range", &range_dataset));
core::ScopedUnref scoped_unref(range_dataset);
DatasetBase* map_dataset;
TF_ASSERT_OK(CreateMapDataset(range_dataset, func_name, &map_dataset)
core::ScopedUnref scoped_unref(map_dataset);
```

jsimsa commented on tensorflow/core/kernels/data/map_dataset_op_test.cc:
```diff
@@ -0,0 +1,363 @@
+ protected:
+  DatasetBase* range_dataset_;
+  DatasetBase* map_dataset_;
+  std::unique_ptr<OpKernel> map_kernel_;
```
I don't think that this needs to be a member variable. The unit tests should define these as local variables.

jsimsa commented on tensorflow/core/kernels/data/map_dataset_op_test.cc:
```diff
@@ -0,0 +1,363 @@
+
+struct DatasetCardinalityTest
+    : MapDatasetOpTest,
+      ::testing::WithParamInterface<std::tuple<int64, int64, int64, int>> {};
```
Use a struct to represent the test case.

jsimsa commented on tensorflow/core/kernels/data/map_dataset_op_test.cc:
```diff
@@ -0,0 +1,363 @@
+struct IteratorRoundtripTest
+    : MapDatasetOpTest,
+      ::testing::WithParamInterface<std::tuple<
+          int64, int64, int64, int, int64, string, std::vector<FunctionDef>>> {
```
ditto

jsimsa commented on tensorflow/core/kernels/data/range_dataset_op_test.cc:
```diff
@@ -30,194 +32,75 @@ namespace tensorflow {
-  std::unique_ptr<FunctionLibraryDefinition> lib_def_;
-  std::unique_ptr<ProcessFunctionLibraryRuntime> pflr_;
-  std::unique_ptr<thread::ThreadPool> thread_pool_;
+  std::unique_ptr<OpKernel> kernel_;
```
do we need any of this state to be member variables or could it just be local variables?

jsimsa commented on tensorflow/core/kernels/data/range_dataset_op_test.cc:
```diff
@@ -342,32 +244,37 @@ struct IteratorRoundtripTest
     }
   }
-  TF_ASSERT_OK(InitSerializationContext());
+  TF_CHECK_OK(CreateSerializationContext(&serialization_context_));
```
TF_ASSERT_OK?

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.cc:
```diff
@@ -0,0 +1,201 @@
+                                   inputs->size(), " vs. ", input_types.size());
+  }
+  bool is_ref = IsRefType(input_types[inputs->size()]);
+  Tensor* input = new Tensor(allocator_, dtype, shape);
```
as per my other suggestion, this should be:

```
std::unique_ptr<Tensor> input = absl::make_unique<Tensor>(allocator_, dtype, shape)

using `inputs->push_back(..., input.get())` and moving `tensors_.push_back(std::move(input))` to be the last action performed by this method
```

jsimsa reviewed (changes requested):
Thanks @feihugis did another pass. I think we are close to being done.

A couple of repeated comments:
- avoid using member variables for things that can be local variables
- use Status consistently as a return value for any function that could fail
- use scoped_unref to properly deallocated refcounted objects
- use named structs to represent test cases for TEST_P
- use smart pointers to manage lifetime of (non-refcounted) objects

feihugis committed:
Address the comments from the third round of review

- Change most of member variables to be local variables

- Refactor a few functions to propagate `Status`

- Use `scope_unref` to deallocate refcounted objects

- Use named structs to represent the test parameters/cases for `TEST_P`

- Use smart pointers to manage lifetime of non-refcounted objects

feihugis commented on tensorflow/core/kernels/data/map_dataset_op_test.cc:
```diff
@@ -88,19 +58,28 @@ class MapDatasetOpTest : public DatasetOpsTestBase {
+    Variant variant = dataset_tensor.scalar<Variant>()();
+    TF_RETURN_IF_ERROR(AddDatasetInputFromArray<Variant>(
+        &map_inputs_, map_kernel->input_types(), TensorShape({}), {variant}));
+    input_dataset->Ref();
```
Add `input_dataset->Ref();` to increase the reference count for `input_dataset`. Otherwise, will get the error: `F ./tensorflow/core/lib/core/refcount.h:102] Check failed: ref_.load() > 0 (0 vs. 0)`.

IIUC, the reason is that `std::vector<std::unique_ptr<Tensor>> tensors_` used for garbage collecting also holds `input_dataset` inside a variant tensor.

feihugis commented:
@jsimsa Thanks for your detailed review! I address the comments in [this commit](https://github.com/feihugis/tensorflow/pull/4/commits/c6e187d86b9ee900d64517d684b63c8a0efdd1e7). As most of member variables are changed to local variables, the code for tests are refactored a lot, but most of them are repeated changes.  Could you help review this commit [c6e187d](https://github.com/feihugis/tensorflow/pull/4/commits/c6e187d86b9ee900d64517d684b63c8a0efdd1e7)?

---
Found two missing `scored_unref`. Added them in the following commit [b398735](https://github.com/feihugis/tensorflow/pull/4/commits/b39873556230e1efc0e77feccb297c64611e18f0).

feihugis committed:
Add two missing

jsimsa reviewed (approved):
This looks great. Once you merge this, let's create a PR against the main TensorFlow repo with your changes so far. Now that you have establish a solid test base, I am comfortable with adding unit tests for other datasets as PR against the TensorFlow repo.

feihugis commented:
@jsimsa Thanks for your help on this PR! I will create a PR against the main TensorFlow repo and work on other unit tests.
